source/geod24/concurrency.d

```diff
@@ -1,1225 +1,765 @@
+        remove, and opApply until it completes.
+
+        Params:
+            rethrow = Rethrow any unhandled exception which may have caused the
+                      current thread to terminate.
+
+        Throws:
+            Any exception not handled by the joined threads.
+
+    ***************************************************************************/
+
+    final void joinAll (bool rethrow = true)
     {
-        try
+        synchronized(this)
         {
-            receiveOnly!string();
-            mainTid.send("");
-        }
-        catch (Throwable th)
-        {
-            mainTid.send(th.msg);
+            // NOTE: This loop relies on the knowledge that m_all uses the
+            //       Thread object for both the key and the mapped value.
+            foreach (Thread t; m_all.keys)
+            {
+                t.join(rethrow);
+            }
         }
     }
 
-    auto tid = spawn(&t1, thisTid);
-    tid.send(1);
-    string result = receiveOnly!string();
-    assert(result == "Unexpected message type: expected 'string', got 'int'");
 }
 
-/**
- * Tries to receive but will give up if no matches arrive within duration.
- * Won't wait at all if provided $(REF Duration, core,time) is negative.
- *
- * Same as `receive` except that rather than wait forever for a message,
- * it waits until either it receives a message or the given
- * $(REF Duration, core,time) has passed. It returns `true` if it received a
- * message and `false` if it timed out waiting for one.
- */
-bool receiveTimeout(T...)(Duration duration, T ops)
-in
-{
-    assert(thisInfo.ident.mbox !is null,
-        "Cannot receive a message until a thread was spawned or thisTid was passed to a running thread.");
-}
-do
-{
-    checkops(ops);
 
-    return thisInfo.ident.mbox.get(duration, ops);
-}
+/*******************************************************************************
 
-@safe unittest
-{
-    static assert(__traits(compiles, {
-        receiveTimeout(msecs(0), (Variant x) {});
-        receiveTimeout(msecs(0), (int x) {}, (Variant x) {});
-    }));
-
-    static assert(!__traits(compiles, {
-        receiveTimeout(msecs(0), (Variant x) {}, (int x) {});
-    }));
-
-    static assert(!__traits(compiles, {
-        receiveTimeout(msecs(0), (int x) {}, (int x) {});
-    }));
-
-    static assert(__traits(compiles, {
-        receiveTimeout(msecs(10), (int x) {}, (Variant x) {});
-    }));
-}
+    Iteratively joins all tracked threads.  This function will block add,
+    remove, and opApply until it completes.
 
-// MessageBox Limits
+*******************************************************************************/
 
-/**
- * These behaviors may be specified when a mailbox is full.
- */
-enum OnCrowding
+public void joinAllThread ()
 {
-    block, /// Wait until room is available.
-    throwException, /// Throw a MailboxFull exception.
-    ignore /// Abort the send and return.
+    ThreadScheduler.instance.joinAll();
```

Comment:
One of the point of the alternative was to remove the global instances. So having a global instance here does not achieve that purpose.